src/logic/pendingStrategy.ts

```diff
@@ -0,0 +1,210 @@
+  dataDecoded: Decode_Data;
+}
+
+interface Decode_Parameter {
+  name: string;
+  type: string,
+  value: string,
+  valueDecoded: Decode_Value[] | Decode_Data;
+}
+
+interface Decode_Data {
+  method: string;
+  parameters: Decode_Parameter[];
+}
+
+interface Decode_Node {
+  data: Decode_Data | Decode_Parameter | Decode_Value;
+  parent: Decode_Node,
+}
+
+class PendingStrategy {
+  transactionHash : string;
+  safeAddresses : string[];
+  bracketAddresses : string[];
+  prices: Decimal[];
+  nonce: number;
+  baseTokenId : number;
+  quoteTokenId : number;
+  baseTokenAddress : string;
+  quoteTokenAddress : string;
+  baseTokenDetails : TokenDetails;
+  quoteTokenDetails : TokenDetails;
+  baseFunding : BN;
+  quoteFunding : BN;
+  owner : string;
+  created : Date;
+  block : any;
+  transactionData : Decode_Data;
+
+  constructor(pendingStrategyTransaction : any) {
+    this.transactionHash = pendingStrategyTransaction.safeTxHash;
+    this.transactionData = pendingStrategyTransaction.dataDecoded;
+    this.nonce = pendingStrategyTransaction.nonce;
+    this.created = new Date(pendingStrategyTransaction.submissionDate);
+  }
+
+  findParamsInTransactionData(transactionData : Decode_Data) : void {
+    let sumFundingTokenBase = toBN(0);
+    let sumFundingTokenQuote = toBN(0);
+
+    let tokenIdBase;
+    let tokenIdQuote;
+
+    let prices = [];
+
+    let bracketAddresses = [];
+
+    if (transactionData.method !== "multiSend") {
+      throw new Error("Expected multiSend to handle pending transaction walking");
+    }
+
+    // let depth = 0
+
+    // This function is responsible for "walking" down the transaction graph.
+    // The reason I consider the structure a graph is because of mutliSend there can be
+    // potentially extremely deeply nested transactions. Decode_Node includes a parent ref
+    // to traverse the tree up and down.
+    let walkTransaction = (node : Decode_Node) : void => {
+      // depth++;
+      //console.log(`${"  ".repeat(depth)}==> Walking "${txData.method}" with ${txData.parameters.length} params`)
+      const txData = node.data as Decode_Data;
+
+      if (txData.method === "placeOrder") {
+        // logger.log('Reading decoded placeOrder', txData);
+        // In placeOrder events we will find the buyToken/sellToken, the buy/sellAmounts (prices)
+        // and if we collect the `to` addresses, the bracket addresses.
+
+        // Find buy and sell token. The first occurrences will be used to determine base/quote
+        // FIXME: I don't know how reliable this will be - it's an assumption that might be wrong
+        //        It might be possible to compare this with current prices for this token pair
+        //        and to do a sanity check if the prices are clsoe enough?
+        const tokenIdBuy = find(txData.parameters, { name: 'buyToken' })?.value;
+        const tokenIdSell = find(txData.parameters, { name: 'sellToken' })?.value;
+        if (tokenIdBase == null && tokenIdQuote == null) {
```

Comment:
Won't the type of `tokenIdBase` and `tokenIdQuote` be `undefined` instead of `null`?

Reply:
`== null` is checking for both `undefined` and `null` and nothing else. I suppose it will always be undefined instead though.
![image](https://user-images.githubusercontent.com/969493/94113603-722e6e00-fe47-11ea-9a22-c35ab99958a3.png)

Reply:
Ah, probably because of double equals instead of triple equals https://stackoverflow.com/a/359509/1272513

I personally prefer to use the strict equal to be precise in regards to what type it's comparing against.